src/backend/executor/exchange_hash_join_executor.h

```diff
@@ -0,0 +1,163 @@
+#pragma once
+
+#include <deque>
+#include <vector>
+
+#include "backend/executor/abstract_join_executor.h"
+#include "backend/executor/exchange_hash_executor.h"
+#include "backend/executor/hash_executor.h"
+#include "backend/planner/hash_join_plan.h"
+
+#include <atomic>
+#include "backend/common/barrier.h"
+#include "backend/common/thread_manager.h"
+#include "boost/lockfree/queue.hpp"
+
+namespace peloton {
+namespace executor {
+
+class ConcurrentOidSet {
+ public:
+  ConcurrentOidSet() {}
+
+  ConcurrentOidSet(ConcurrentOidSet &&that)
+      : container_(std::move(that.container_)) {}
+  ConcurrentOidSet(const ConcurrentOidSet &that) = delete;
+
+  ConcurrentOidSet &operator=(const ConcurrentOidSet &that) = delete;
+  ConcurrentOidSet &operator=(ConcurrentOidSet &&that) = delete;
+
+  size_t MUTEX_CNT = 16;
+  std::array<std::mutex, 16> mtx_list_;
+  std::unordered_set<oid_t> container_;
+
+  void Erase(oid_t key) {
+    int mutex_id = key % MUTEX_CNT;
+    mtx_list_[mutex_id].lock();
+    container_.erase(key);
+    mtx_list_[mutex_id].unlock();
+  }
+
+  bool Empty() { return container_.empty(); }
+};
+
+typedef std::uint_least32_t thread_no;
+
+class PesudoBarrier {
+ public:
+  void Release() {
+    std::lock_guard<std::mutex> lock(mutex_);
+    ++count_;
+    assert(count_ <= total_);
+    if (count_ == total_) {
+      done = true;
+    }
+  }
+  PesudoBarrier() : total_(0) {}
+  void SetTotal(thread_no new_total) { total_ = new_total; }
+  bool IsDone() { return done; }
+  bool IsNoNeedToDo() { return no_need_to_do_; }
+  void SetNeedToDo(bool value) { no_need_to_do_ = value; }
+
+ private:
+  thread_no total_;
+  bool done = false;
+  std::mutex mutex_;
+  size_t count_ = 0;
+  bool no_need_to_do_ = false;
+};
+
+typedef std::vector<ConcurrentOidSet> ExHashJoinRowSets;
+
+class ExchangeHashJoinExecutor : public AbstractJoinExecutor {
+  ExchangeHashJoinExecutor(const ExchangeHashJoinExecutor &) = delete;
+  ExchangeHashJoinExecutor &operator=(const ExchangeHashJoinExecutor &) =
+      delete;
+
+ public:
+  explicit ExchangeHashJoinExecutor(const planner::AbstractPlan *node,
+                                    ExecutorContext *executor_context);
+
+  std::vector<LogicalTile *> GetOutputs();
+
+  ~ExchangeHashJoinExecutor() = default;
+
+  void GetRightHashTable(Barrier *barrier);
+  void GetLeftScanResult(Barrier *barrier);
+
+  void Probe(std::atomic<thread_no> *no, PesudoBarrier *barrier);
+  void UpdateLeftJoinRowSets();
+  void UpdateRightJoinRowSets();
+
+  // helper function to launch number worker threads, each of which will do
+  // function
+  static void LaunchWorkerThreads(size_t number,
+                                  std::function<void()> function) {
+    LOG_TRACE("LaunchWorkerThreads(%lu)", number);
```

Comment:
Thanks.
Actually this method will not launch any threads. Conceptually, it will submit 'number' of threads, each of which will execute 'function'. However, in fact, it only submits 'number' tasks to the thread pool.
Maybe we need to make this clearer in the function name.